ygen/schemaparse.go

```diff
@@ -17,9 +17,79 @@ package ygen
 import (
 	"bytes"
 	"compress/gzip"
+	"encoding/json"
 	"fmt"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
 )
 
+// buildJSON tree takes an input set of modules expressed as a slice of yang.Entry
+// structs, the names of directories within the generated code, and a yang.Entry
+// which corresponds to the fake root and returns a JSON serialised tree of the
+// schema for the set of modules. The JSON document that is returned is always
+// rooted on a yang.Entry which corresponds to the root item, and stores all
+// root-level enties (and their subtrees) within the input module set. All
+// YANG directories are annotated in the output JSON with the name of the type
+// they correspond to in the generated code, and the absolute schema path that
+// the entry corresponds to.
+func buildJSONTree(ms []*yang.Entry, dn map[string]string, fakeroot *yang.Entry) ([]byte, error) {
+	rootEntry := &yang.Entry{
+		Dir:        map[string]*yang.Entry{},
+		Annotation: map[string]interface{}{},
+	}
+	for _, m := range ms {
+		annotateChildren(m, dn)
+		for _, ch := range children(m) {
+			if _, ex := rootEntry.Dir[ch.Name]; ex {
+				return nil, fmt.Errorf("overlapping root children for key %s", ch.Name)
+			}
+			rootEntry.Dir[ch.Name] = ch
+		}
+	}
+
+	if fakeroot != nil {
+		rootEntry.Name = fakeroot.Name
+		rootEntry.Annotation["schemapath"] = "/"
+		rootEntry.Annotation["structname"] = dn[fakeroot.Path()]
+		rootEntry.Kind = yang.DirectoryEntry
+	}
+	// Always annotate the root as a fake root, so that it is not treated
+	// as a path element in ytypes.
+	rootEntry.Annotation["isFakeRoot"] = true
+
+	j, err := json.MarshalIndent(rootEntry, "", strings.Repeat(" ", 4))
+	if err != nil {
+		return nil, fmt.Errorf("JSON marshalling error: %v", err)
+	}
+	return j, nil
+}
+
+// annotateChildren annotates e with its schema path, and the value corresponding
+// to its path in the supplied dn map. The dn map is assumed to contain the
+// names of unique directories that are generated within the code to be output.
+// The children of e are recursively annotated.
+func annotateChildren(e *yang.Entry, dn map[string]string) {
+	for _, ch := range children(e) {
+		if ch.Annotation == nil {
+			ch.Annotation = map[string]interface{}{}
+		}
+		if n, ok := dn[ch.Path()]; ok {
+			ch.Annotation["structname"] = n
+		}
+		if ch.IsDir() {
+			ch.Annotation["schemapath"] = ch.Path()
+			// Recurse to annotate the children of this entry.
+			annotateChildren(ch, dn)
+		}
+		// In the compressed schema, we remove the human-readable Description
+		// of each child element, since this is not required by any machine
+		// reconstructing the data.
+		ch.Description = ""
+	}
+	e.Description = ""
```

Comment:
e itself doesn't seem to get annotated?

Reply:
Fixed this - with the current logic it's not actually required since we always call this with a module entry, which isn't included in the tree, so necessarily all containers get annotated, but just in case we change, and to make this consistent I fixed it.